xdsl/printer.py

```diff
@@ -534,10 +536,27 @@ def print_attribute(self, attribute: Attribute) -> None:
 
         if isinstance(attribute, FloatAttr):
             value = attribute.value
-            attr_type = cast(
-                FloatAttr[Float16Type | Float32Type | Float64Type], attribute
-            ).type
-            self.print_string(f"{value.data:.6e} : ")
+            attr_type = cast(AnyFloatAttr, attribute).type
+            if math.isnan(value.data) or math.isinf(value.data):
+                if isinstance(attr_type, Float16Type):
+                    self.print_string(
+                        f"{hex(struct.unpack('<H', struct.pack('<e', value.data))[0])} : "
+                    )
+                elif isinstance(attr_type, Float32Type):
+                    self.print_string(
+                        f"{hex(struct.unpack('<I', struct.pack('<f', value.data))[0])} : "
+                    )
+                elif isinstance(attr_type, Float64Type):
+                    self.print_string(
+                        f"{hex(struct.unpack('<Q', struct.pack('<d', value.data))[0])} : "
+                    )
```

Comment:
Now that I think about it, it feels like it would be better to use the helpers in `bitwise_casts` after all, using the struct or ctypes approach. We will need the logic in a few places, such as when storing float data in assembly, and other optimisations, so it would be worth having it in an easily reusable place.

Reply:
Done as suggested.